src/twisted/test/test_defer.py

```diff
@@ -1735,6 +1736,37 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
         for thing in thingsThatAreNotCoroutines:
             self.assertRaises(defer.NotACoroutineError, Deferred.fromCoroutine, thing)
 
+    def test_inlineCallbacksCancelCaptured(self) -> None:
+        """
+        Cancelling an `inlineCallbacks` correctly handles the function catching
+        the `CancelledError`.
+        """
+        canceller1Calls: List[Deferred[Any]] = []
+        canceller2Calls: List[Deferred[Any]] = []
+        d1: Deferred[Any] = Deferred(canceller1Calls.append)
+        d2: Deferred[Any] = Deferred(canceller2Calls.append)
+
+        @defer.inlineCallbacks
+        def testFunc() -> Any:
+            try:
+                yield d1
+            except Exception:
+                pass
+
+            yield d2
+
+        funcD = testFunc()
+        self.assertFalse(funcD.called)
+
+        funcD.cancel()
```

Comment:
Maybe add a comment explaining the correct behaviour
```suggestion
        # On the first cancel it just handled by the
        # function by ignoring the exception.
        # The function still has no results as the
        # second deferred is not yet called.
```